File: backend/schemas/server_import.py
# backend/schemas/server_import.py
from datetime import datetime
from typing import Literal

from pydantic import Field

from schemas.common import CamelModel

# --- Browse ---


class ServerConnectRequest(CamelModel):
    """Connect to an FTP/SFTP server and list a directory."""

    protocol: Literal["ftp", "sftp"]
    host: str = Field(min_length=1, max_length=255)
    port: int | None = None
    username: str = Field(min_length=1, max_length=255)
    password: str = Field(min_length=1)
    path: str = "/"
    saved_server_id: int | None = None  # if connecting via saved server


class RemoteFileEntry(CamelModel):
    """A file or directory on the remote server."""

    name: str
    path: str
    is_dir: bool
    size: int | None = None


class ServerBrowseResponse(CamelModel):
    """Directory listing from a remote server."""

    current_path: str
    entries: list[RemoteFileEntry]


# --- Import ---


class ServerImportRequest(CamelModel):
    """Import specific FASTQ files from a remote server."""

    protocol: Literal["ftp", "sftp"]
    host: str = Field(min_length=1, max_length=255)
    port: int | None = None
    username: str = Field(min_length=1, max_length=255)
    password: str = Field(min_length=1)
    file_paths: list[str] = Field(min_length=1)
    save_server: bool = False
    server_name: str | None = None  # required if save_server is True


class ServerImportStartedResponse(CamelModel):
    """Returned after starting a background import."""

    import_id: str
    file_count: int
    message: str


# --- Progress ---


class ImportFileProgress(CamelModel):
    """Progress for a single file in an import."""

    remote_path: str
    filename: str
    status: Literal["pending", "downloading", "complete", "error"]
    bytes_downloaded: int = 0
    bytes_total: int | None = None
    error: str | None = None


class ServerImportProgress(CamelModel):
    """Overall import progress."""

    import_id: str
    experiment_id: int
    user_id: int
    status: Literal["connecting", "downloading", "complete", "error"]
    files: list[ImportFileProgress]
    completed_count: int
    total_count: int
    error: str | None = None


# --- Saved Servers ---


class SavedServerCreate(CamelModel):
    """Create a new saved server connection."""

    name: str = Field(min_length=1, max_length=100)
    protocol: Literal["ftp", "sftp"]
    host: str = Field(min_length=1, max_length=255)
    port: int | None = None
    username: str = Field(min_length=1, max_length=255)
    password: str = Field(min_length=1)
    default_path: str = "/"


class SavedServerUpdate(CamelModel):
    """Update a saved server — all fields optional."""

    name: str | None = Field(default=None, min_length=1, max_length=100)
    default_path: str | None = None
    username: str | None = Field(default=None, min_length=1, max_length=255)
    password: str | None = Field(default=None, min_length=1)


class SavedServerRead(CamelModel):
    """Saved server response — password is NEVER included."""

    id: int
    name: str
    protocol: str
    host: str
    port: int | None
    username: str
    default_path: str
    created_at: datetime
    updated_at: datetime

File: frontend/src/api/serverImport.ts
// frontend/src/api/serverImport.ts
import client from './client';

// --- Types ---

export interface ServerConnectRequest {
  protocol: 'ftp' | 'sftp';
  host: string;
  port?: number;
  username: string;
  password: string;
  path: string;
  savedServerId?: number;
}

export interface RemoteFileEntry {
  name: string;
  path: string;
  isDir: boolean;
  size: number | null;
}

export interface ServerBrowseResponse {
  currentPath: string;
  entries: RemoteFileEntry[];
}

export interface ServerImportRequest {
  protocol: 'ftp' | 'sftp';
  host: string;
  port?: number;
  username: string;
  password: string;
  filePaths: string[];
  saveServer?: boolean;
  serverName?: string;
}

export interface ServerImportStartedResponse {
  importId: string;
  fileCount: number;
  message: string;
}

export interface ImportFileProgress {
  remotePath: string;
  filename: string;
  status: 'pending' | 'downloading' | 'complete' | 'error';
  bytesDownloaded: number;
  bytesTotal: number | null;
  error: string | null;
}

export interface ServerImportProgress {
  importId: string;
  experimentId: number;
  userId: number;
  status: 'connecting' | 'downloading' | 'complete' | 'error';
  files: ImportFileProgress[];
  completedCount: number;
  totalCount: number;
  error: string | null;
}

export interface SavedServer {
  id: number;
  name: string;
  protocol: 'ftp' | 'sftp';
  host: string;
  port: number | null;
  username: string;
  defaultPath: string;
  createdAt: string;
  updatedAt: string;
}

export interface SavedServerCreate {
  name: string;
  protocol: 'ftp' | 'sftp';
  host: string;
  port?: number;
  username: string;
  password: string;
  defaultPath?: string;
}

// --- API Functions ---

export async function browseServer(
  experimentId: number,
  req: ServerConnectRequest,
): Promise<ServerBrowseResponse> {
  const { data } = await client.post<ServerBrowseResponse>(
    `/experiments/${experimentId}/server-import/browse`,
    req,
  );
  return data;
}

export async function startImport(
  experimentId: number,
  req: ServerImportRequest,
): Promise<ServerImportStartedResponse> {
  const { data } = await client.post<ServerImportStartedResponse>(
    `/experiments/${experimentId}/server-import/start`,
    req,
  );
  return data;
}

export async function getImportProgress(
  experimentId: number,
  importId: string,
): Promise<ServerImportProgress> {
  const { data } = await client.get<ServerImportProgress>(
    `/experiments/${experimentId}/server-import/${importId}/progress`,
  );
  return data;
}

export async function listSavedServers(): Promise<SavedServer[]> {
  const { data } = await client.get<SavedServer[]>('/users/me/saved-servers');
  return data;
}

export async function createSavedServer(body: SavedServerCreate): Promise<SavedServer> {
  const { data } = await client.post<SavedServer>('/users/me/saved-servers', body);
  return data;
}

export async function deleteSavedServer(serverId: number): Promise<void> {
  await client.delete(`/users/me/saved-servers/${serverId}`);
}
